src/main/scala/workflow/AutoCacheRule.scala

```diff
@@ -0,0 +1,531 @@
+
+    val toCache = cached.intersect(dataOutputtingInstructions)
+
+    pipe.indices.foldLeft (
+      (Seq[Instruction](), pipe.indices.zipWithIndex.toMap + (Pipeline.SOURCE -> Pipeline.SOURCE))
+    ) {
+      case ((newPipe, oldToNewIndexMap), i) if toCache.contains(i) =>
+        (newPipe ++ Seq(
+          pipe(i).mapDependencies(oldToNewIndexMap),
+          new Cacher,
+          TransformerApplyNode(oldToNewIndexMap(i) + 1, Seq(oldToNewIndexMap(i)))
+        ),
+          oldToNewIndexMap.map {
+            case (key, value) => if (key >= i) (key, value + 2) else (key, value)
+          })
+
+      case ((newPipe, oldToNewIndexMap), i) =>
+        (newPipe :+ pipe(i).mapDependencies(oldToNewIndexMap), oldToNewIndexMap)
+    }._1
+  }
+
+  def aggressiveCache(instructions: Seq[Instruction]): Seq[Instruction] = {
+    val immediateChildren = getImmediateChildrenByInstruction(instructions)
+    val nodeWeights = getNodeWeights(instructions)
+
+    val childrenOfSource = WorkflowUtils.getChildren(Pipeline.SOURCE, instructions)
+
+    // Cache any node whose direct output is used more than once while training
+    val instructionsToCache = instructions.indices.filter {
+      i => immediateChildren(i).filterNot(childrenOfSource).map(nodeWeights).sum > 1
+    }.toSet
+
+    makeCachedPipeline(instructions, instructionsToCache)
+  }
+
+  def cacheMem(caches: Set[Int], profiles: Map[Int, Profile]): Long = {
+    // Must do a toSeq here otherwise the map may merge equal memories
+    caches.toSeq.map(i => profiles.getOrElse(i, Profile(0, 0, 0)).rddMem).sum
+  }
+
+  /**
+   * Returns true iff there is still an uncached node whose output is used > once, that would fit in memory
+   * if cached
+   */
+  def stillRoom(caches: Set[Int], runs: Map[Int, Int], profiles: Map[Int, Profile], spaceLeft: Long): Boolean = {
+    runs.exists { i =>
+      (i._2 > 1) &&
+        (!caches.contains(i._1)) &&
+        (profiles.getOrElse(i._1, Profile(0, 0, 0)).rddMem < spaceLeft)
+    }
+  }
+
+  def selectNext(
+    pipe: Seq[Instruction],
+    profiles: Map[Int, Profile],
+    cached: Set[Int],
+    runs: Map[Int, Int],
+    spaceLeft: Long
+  ): Int = {
+    //Get the uncached node which fits that maximizes savings in runtime.
+    pipe.indices.filter(i =>
+      !cached(i) &&
+        profiles.getOrElse(i, Profile(0, 0, 0)).rddMem < spaceLeft &&
+        runs(i) > 1)
+      .minBy(i => estimateCachedRunTime(pipe, cached + i, profiles))
+  }
+
+  def greedyCache(
+    instructions: Seq[Instruction],
+    profiles: Map[Int, Profile],
+    maxMem: Option[Long]
+  ): Seq[Instruction] = {
+    val nodeWeights = getNodeWeights(instructions)
+
+    var cached = initCacheSet(instructions)
+    val memBudget = maxMem.getOrElse(instructions.collectFirst {
+      case SourceNode(rdd) =>
+        val sc = rdd.sparkContext
+        // Calculate memory budget, ignoring driver unless in local mode
+        val workers = if (sc.getExecutorStorageStatus.size == 1) {
```

Comment:
With `local[4]`, getExecutorStorageStatus returns a single executor status, the driver's. This would go down the first code path and return the correct memory budget.